index.js

```diff
@@ -406,54 +406,41 @@ class Hyperbee extends ReadyResource {
 
   watch (range, opts) {
     if (!this._watchers) throw new Error('Can only watch the main bee instance')
-    return new Watcher(this, range, opts)
+    return new RangeWatchStream(this, range, opts)
   }
 
-  async getAndWatch (key, opts) {
+  getAndWatch (key, opts) {
     if (!this._watchers) throw new Error('Can only watch the main bee instance')
-
-    const watcher = new EntryWatcher(this, key, opts)
-    await watcher._debouncedUpdate()
-
-    if (this.closing) {
-      await watcher.close()
-      throw new Error('Bee closed')
-    }
-
-    return watcher
+    return new KeyWatchStream(this, key, opts)
   }
 
   _onappend () {
     for (const watcher of this._watchers) {
       watcher._onappend()
     }
-
-    for (const watcher of this._entryWatchers) {
-      watcher._onappend()
-    }
   }
 
   _ontruncate () {
     for (const watcher of this._watchers) {
       watcher._ontruncate()
     }
-
-    for (const watcher of this._entryWatchers) {
-      watcher._ontruncate()
-    }
   }
 
   _makeSnapshot () {
     // TODO: better if we could encapsulate this in hypercore in the future
     return this._checkout <= this.core.length ? this.core.snapshot() : this.core.session()
   }
 
+  session (opts) {
+    return this.checkout(0, opts)
+  }
+
   checkout (version, opts = {}) {
     // same as above, just checkout isn't set yet...
 
     const snap = opts.reuseSession
       ? this
-      : version <= this.core.length ? this.core.snapshot() : this.core.session()
+      : (version > 0 && version <= this.core.length) ? this.core.snapshot() : this.core.session()
```

Comment:
Whether or not that's smart is a diff discussion 😆

Reply:
Aha, hadn't followed the flow that far, in that case the only difference is on calling `this.core.snapshot()` instead of `this.core.session()`, where I'm not sure if it has any practical impact

Reply:
but what we can do is make checkout coerce 0 to 1 outside the session usage, thats prop a good idea

Reply:
fixed